File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        vector<int> indexVec;
        indexVec.push_back(-100000);
        int minDist = 100000;
        
        int before = head->val;
        head = head->next;
        
        int i = 1;
        while(head->next != NULL) {
            if(max(before, head->next->val) < head->val || min(before, head->next->val) > head->val) {
                minDist = min(minDist, i - indexVec.back());
                indexVec.push_back(i);
            }
            i++;
            before = head->val;
            head = head->next;
        }
        
        if(indexVec.size() <= 2) return vector<int>({-1, -1});
        return vector<int>({minDist, indexVec.back() - indexVec[1]});
    }
};

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        before = head.val
        head = head.next
        minDist = 100000
        indexList = [-100000]
        
        i = 1
        while head.next != None:
            if head.val < min(before, head.next.val) or max(before, head.next.val) < head.val:
                minDist = min(minDist, i - indexList[-1])
                indexList.append(i)
            i += 1
            before = head.val
            head = head.next
        
        return [minDist, indexList[-1] - indexList[1]] if len(indexList) > 2 else [-1, -1];